src/modules/expenses/expense.handlers.ts

```diff
@@ -0,0 +1,126 @@
+import type { AppRouteHandler } from "@/common/lib/types";
+
+import { ActivityType, AuthRoles } from "@/common/enums";
+import { logActivity } from "@/common/helpers/activity-log.helper";
+import * as HTTPStatusCodes from "@/common/utils/http-status-codes.util";
+
+import type { TCreateExpenseRoute } from "./expense.routes";
+
+import { getCategoryRepository } from "../categories/category.repository";
+import {
+  createExpenseRepositoryViaAdmin,
+  createExpenseRepositoryViaUser,
+} from "./expense.repository";
+
+export const createExpense: AppRouteHandler<TCreateExpenseRoute> = async (
+  c,
+) => {
+  const logger = c.get("logger");
+  const user = c.get("user");
+  const payload = c.req.valid("json");
+
+  if (!user) {
+    logger.debug("User is not authorized to create expense");
+    return c.json(
+      {
+        success: false,
+        message: "You are not authorized, please login",
+      },
+      HTTPStatusCodes.UNAUTHORIZED,
+    );
+  }
+
+  let expense;
+  switch (user.role) {
+    case AuthRoles.USER: {
+      const { categoryId } = payload;
+
+      // check if category either belongs to user or global
+      if (categoryId) {
+        const category = await getCategoryRepository(categoryId);
+
+        if (!category.userId && category.userId !== user.id) {
+          logger.debug("Category does not belongs to user");
+          return c.json(
+            {
+              success: false,
+              message: "Category does not belongs to user",
+            },
+            HTTPStatusCodes.BAD_REQUEST,
+          );
+        }
+      }
+
+      expense = await createExpenseRepositoryViaUser(payload, user?.id);
+      break;
+    }
+    case AuthRoles.ADMIN: {
+      const { payerId, categoryId } = payload;
+      if (!payerId) {
+        logger.debug("Missing payer Id to create expense");
```

Comment:
This is not quite right(according to me), we have to do it in a way such that anyone can send the payerId, if they do send it use that or else default it to the logged in user id.
Please let me know your thoughts about this approach.

Reply:
Related to https://github.com/GenieWizards/finance-management-api/pull/68#discussion_r1864188154.  Replied.